### ISSUE TITLE
Squashing commits on merge

### ISSUE THREAD

**Issue by Y0hy0h** [mswart/openmensa-parsers#68]
I'd like to make an argument against squashing the commits on merge:

Squashing the commits makes it hard to identify bugs that were introduced in a PR. If we notice a bug on master, and say we narrow it down to commit 5dd40a44f87e5febb635409cb1d82d98af8d1aff that introduced it, then that doesn't help us much. We would need to somehow go through the PR's commits to narrow it down appropriately.

The only reason I can imagine that speaks _for_ squashing the commits, is that the history is "cleaner" in the sense that it is easier to figure out what commits pertain to a PR, since they all get squashed into one.  
But if you have a tree view of the commits, is very easy to identify which belong to a PR, but it allows to narrow down bugs much more easily.
![image](https://user-images.githubusercontent.com/11377826/34774997-dfef1cd2-f611-11e7-9575-12ddb5923df4.png)
I have to admit that the tree view gets a little bit ugly when a PR has to handle merge conflicts with master during its lifetime, but the benefit of being able to track down bugs and see exactly what happened is very important to me.

What do you think?

**Comment by klemens** [comment]
> I have to admit that the tree view gets a little bit ugly when a PR has to handle merge conflicts with master during its lifetime

Which is why you should avoid merging master into a pr-branch and instead rebase the branch onto the current master.

Specifically in #66 there were many unnecessary commits, which changed the same thing multiple times or were reverted in the same branch, which really makes it hard to follow if you are just reading though all master commits. I suggest that you squash such commits and reorder them into a logical structure before finishing the pr (can be done really easily using git rebase --interactive).

> Would it have been a good idea to rebase this PR? Because the commits were already published, a rebase would change all their hashes which could potentially lead to trouble [quote from #63]

Changing commit ids is fine as long as they are not in a branch that others work with like master. In your own pr-branch, you can rewrite history however you like. 😉

**Comment by Y0hy0h** [comment]
I am very hesitant about rebasing published commits, because of arguments such as in the [git documentation](https://git-scm.com/book/en/v1/Git-Branching-Rebasing#The-Perils-of-Rebasing). If anyone had fetched the PR, then the rebase would have messed up their local repo.  
That's why I by default opt to merge the changes from master instead of rebasing (which I would have found nicer, too!).

#66 is an extreme example, because I couldn't really test the build locally. So I had to push all these experimental commits. I totally get your point, but I'm still very hesitant about rebasing published commits.

But I guess, realistically it comes down to preference of the repository contributors. If we explicitely allow rebases to PRs, then we might discourage direct contributions to PRs. But we could handle issues such as in #66.

Apart from the rebasing issue, what do you think of the squashing question? :)

**Comment by mswart** [comment]
> Squashing the commits makes it hard to identify bugs that were introduced in a PR. If we notice a bug on master, and say we narrow it down to commit 5dd40a4 that introduced it, then that doesn't help us much. We would need to somehow go through the PR's commits to narrow it down appropriately.

The commit message clearly says that it is PR #66. So why would it be difficult?
I my experience having multiple commits changing stuff over and over is the real challenge to identify how a bug was introduced.

During feature development I understand the need for many often small changesets. But dealing with all these small adjustments and learning steps is often struggling afterwards. Learning is not a straight way.
So in the end it is good to condense the knowledge into few larger commits. This allows also for more descriptive commit message: a ideal commit messages explains why a change was made? Did other approaches failed? Why? The code changes of a commit is tracked by git. But the motivation is often lost. Reverting any of these resulting larger commits should still result in a mostly working repository.

A bug or change is traced back to a commit: the commit messages helps to decide how to proceed who? It is safe to revert the commit as the motivation behind the change is not longer true? Is is now possible to choose a different way to solve the issue: maybe the dependencies didn't allow a different solution.

In regards #66: I thought about it is reasonable to split the PR into multiple commits dealing with different aspects. But the PR simply moves files around. Additional commits for move the `parse` method and 
improve thing the `setup.py` might be reasonable. But in my opinion creating these different commits would have taken much more time than the gained effect.

> I am very hesitant about rebasing published commits, because of arguments such as in the git documentation. If anyone had fetched the PR, then the rebase would have messed up their local repo.
That's why I by default opt to merge the changes from master instead of rebasing (which I would have found nicer, too!).

But few people fetch/merge foreign PR branches. In most cases they know what they are doing and can deal with such situations or they work with their own history.
Overall the advantages outweigh the possible disadvantages.

In addition it is also common to delay rebasing until the merge. When the changeset itself accepted, all commits are rebase and restructured and afterwards merged directly.

**Comment by Y0hy0h** [comment]
Ok, I think the bottom line is that rebasing the PR for clarity is a good idea and squashing commits on merge is the way to go here.

I'm still not convinced that losing the individual commits is a good idea. But it boils down to the following question: How would you checkout e. g. 0a6c76e40038dc334b19f314f8bce6ff2cc61067 from #65? Maybe there is an elegant way that I don't know of. And this is just an issue I see that I would love to see a solution to. ;)

With this discussion, I only wanted to learn about the reasons you chose to squash, because I only had arguments against it. I don't want to insist on anything, and I will totally handle this the way that everyone agrees it should be handled (rebase + squash). :)

**Comment by mswart** [comment]
Why would I want to checkout 0a6c76e40038dc334b19f314f8bce6ff2cc61067 from #65? I mean it is textual documentation. I can hardly think of any situation where to you want to revert a specific change set with in PR.

After the rebasing and squashing the individual commits should be enough to handle feature adjustments. If specific details needs to change, you would probably create a new commit nevertheless.

**Comment by Y0hy0h** [comment]
I have done some debugging in the past where I would figure out what commit introduced the bug. Since I aim to commit very often and only very small changes, this makes it much easier to figure out where the bug is coming from, since it comes from a very concise code change.

I shouldn't have picked #65, because it doesn't fit into that example. However, because I realized that I basically only want to figure out how to access the individual commits, I would be satisfied by knowing how to checkout a specific commit from a PR that was squashed.  
Because then, if there was a bug on master, I could first narrow it down to the PR, and then find the commit that introduced the bug.

This is not meant specifically for this repo, this is a general question. I'm just trying to figure out the most elegant way of doing this. ;)

So my question is: How can I checkout 0a6c76e40038dc334b19f314f8bce6ff2cc61067 from #65?  
But not in the sense that I really want to. Just to understand how I can go through the commits of a squashed PR.

If that is not easily possible, then I'll just have to deal with PRs being squashed. :grin:

I hope this clarifies my question. :D

**Comment by azrdev** [comment]
> how to checkout a specific commit from a PR that was squashed.

you cannot, git creates a new squash commit, and as soon as the original branch of the PR is lost there is no way to get its commits back

**Comment by Y0hy0h** [comment]
Thanks, @azrdev! That's what I feared...

In my opinion, it's worth keeping a PR's commits and not squashing them. But I do realize now that I have a bit more freedom rebasing PRs, since most are not cloned in practice.

I'll just accept that PRs are squashed on this project. :D
